### PR TITLE
WIP: changes to Cube data api

### PR DESCRIPTION
**Pull request by dkillick**
Specifically, try and use the data setter for _every_ change to `cube.data`. This includes:

* instantiating a cube,
* realising lazy data, and
* user changing a cube's data.

This set of changes started life in #2451 but got backed out of there when it got big and gnarly.

### CODE REVIEW

File: lib/iris/_lazy_data.py [hunk 1]
```diff
@@ -97,7 +97,10 @@ def as_concrete_data(data, **kwargs):
 
     """
     if is_lazy_data(data):
-        # Realise dask array.
+        # Realise dask array, ensuring the data result is always a NumPy array.
+        # In some cases dask may return a scalar numpy.int/numpy.float object
+        # rather than a numpy.ndarray object.
+        # Recorded in https://github.com/dask/dask/issues/2111.
```
Comment (dkillick) [inline lib/iris/_lazy_data.py]: This comment should probably be moved now that we're not ensuring `data` is a NumPy array here.

File: lib/iris/cube.py [hunk 2]
```diff
@@ -664,9 +663,9 @@ def __init__(self, data, standard_name=None, long_name=None,
             This object defines the shape of the cube and the phenomenon
             value in each cell.
 
-            It can be a biggus array, a numpy array, a numpy array
-            subclass (such as :class:`numpy.ma.MaskedArray`), or an
-            *array_like* as described in :func:`numpy.asarray`.
+            ``data`` can be a dask array, a NumPy array, a NumPy array
+            subclass (such as :class:`numpy.ma.MaskedArray`), or
+            array_like (as described in :func:`numpy.asarray`).
```
Comment (dkillick) [inline lib/iris/cube.py]: Whoops - looks like this change got missed when I backed this out of #2451...

File: lib/iris/cube.py [hunk 3]
```diff
@@ -1754,8 +1750,16 @@ def data(self):
 
     @data.setter
     def data(self, value):
-        if not (hasattr(value, 'shape') and hasattr(value, 'dtype')):
```
Comment (dkillick) [inline lib/iris/cube.py]: This turns out to not be the complete check of what's required here. For example, a `numpy.int32` object has a `shape` attribute but is **not** an `ndarray`...

**Comment by marqh** [comment]
hi @dkillick 

please may you confirm if this PR is still a desired change?

if so, please rebase as it now conflicts
if not, please close

thank you

**Comment by dkillick** [comment]
Aha, no - this all got done by @bjlittle in the end. Thanks @marqh

**Pull request closed without merging**